smack-extensions/src/main/java/org/jivesoftware/smackx/mood/MoodManager.java

```diff
@@ -0,0 +1,177 @@
+                                listener.onMoodUpdated(contact, mood);
+                            }
+                        }
+                    }, "MoodListener");
+                }
+            }
+        });
+        ServiceDiscoveryManager.getInstanceFor(connection).addFeature(MOOD_NOTIFY);
+    }
+
+    public static MoodManager getInstanceFor(XMPPConnection connection) {
+        MoodManager manager = INSTANCES.get(connection);
+        if (manager == null) {
+            manager = new MoodManager(connection);
+            INSTANCES.put(connection, manager);
+        }
+        return manager;
+    }
+
+    public void setMood(Mood mood)
+            throws InterruptedException, SmackException.NotLoggedInException, SmackException.NoResponseException,
+            SmackException.NotConnectedException, XMPPException.XMPPErrorException, PubSubException.NotALeafNodeException {
+        setMood(mood, null, null);
+    }
+
+    public void setMood(Mood mood, String text)
+            throws InterruptedException, SmackException.NotLoggedInException, SmackException.NoResponseException,
+            SmackException.NotConnectedException, XMPPException.XMPPErrorException, PubSubException.NotALeafNodeException {
+        setMood(mood, null, text);
+    }
+
+    public void setMood(Mood mood, MoodConcretisation concretisation)
+            throws InterruptedException, SmackException.NotLoggedInException, SmackException.NoResponseException,
+            SmackException.NotConnectedException, XMPPException.XMPPErrorException, PubSubException.NotALeafNodeException {
+        setMood(mood, concretisation, null);
+    }
+
+    public void setMood(Mood mood, MoodConcretisation concretisation, String text)
+            throws InterruptedException, SmackException.NotLoggedInException, SmackException.NoResponseException,
+            SmackException.NotConnectedException, XMPPException.XMPPErrorException, PubSubException.NotALeafNodeException {
+        MoodElement element = buildMood(mood, concretisation, text);
+        publishMood(element);
+    }
+
+    public void clearMood()
+            throws InterruptedException, SmackException.NotLoggedInException, SmackException.NoResponseException,
+            SmackException.NotConnectedException, XMPPException.XMPPErrorException, PubSubException.NotALeafNodeException {
+        MoodElement element = buildMood(null, null, null);
+        publishMood(element);
+    }
+
+    private void publishMood(MoodElement moodElement)
+            throws SmackException.NotLoggedInException, InterruptedException, PubSubException.NotALeafNodeException,
+            XMPPException.XMPPErrorException, SmackException.NotConnectedException, SmackException.NoResponseException {
+
+        if (!connection().isAuthenticated()) {
+            throw new SmackException.NotLoggedInException();
+        }
+        PubSubManager pubSubManager = PubSubManager.getInstance(connection(), connection().getUser().asBareJid());
```

Comment:
A `Manager`'s `getInstanceFor()` method is expensive, why not have `MoodManager` hold a reference to `PubSubManager` in  a private field?